File: ev_point/lib/src/features/map/domain/entities/station.dart
// lib/src/features/map/domain/entities/station.dart
class Station {
  final String id;
  final String name;
  final String address;
  final double longitude;
  final double latitude;
  final String connectorType;
  final int powerKw;
  final int totalPoints;
  final int availablePoints;
  final int pricePerKwh;
  final String status;

  bool get available => status.toLowerCase() == 'online' && availablePoints > 0;

  Station({
    required this.id,
    required this.name,
    required this.address,
    required this.longitude,
    required this.latitude,
    required this.connectorType,
    required this.powerKw,
    required this.totalPoints,
    required this.availablePoints,
    required this.pricePerKwh,
    required this.status,
  });
}

File: Station_Management_Service/src/models/Station.js
import mongoose from "mongoose";
// Trạm sạc 
const stationSchema = new mongoose.Schema(
    {
    // Tên trạm
    name: { type: String, required: true, trim: true },
    // Địa chỉ chi tiết
    address: { type: String, trim: true },
    // Toạ độ địa lý
    longitude: { type: Number, required: true },
    latitude: { type: Number, required: true },
    // Loại cổng sạc: CCS, AC_Type2, v.v.
    connector_type: { type: String, trim: true },
    // Công suất kW
    power_rating: { type: Number, default: 0 },
    // Tổng điểm sạc trong trạm
    total_points: { type: Number, default: 0 },
    // Số điểm còn trống
    available_points: { type: Number, default: 0 },
    // Giá điện / kWh
    price_per_kwh: { type: Number, default: 0 },
    // Trạng thái trạm
    status: {type: String,enum: ["online", "offline", "maintenance"],default: "online",},
    location: {
        type:{type:String,enum:["Point"],default:"Point"},
        coordinates: {type: [Number], required: true,},
    }
    },
    { timestamps: true }
)
stationSchema.index({ location: "2dsphere" });
// ✅ dùng pre('validate') thay vì pre('save')
stationSchema.pre("validate", function (next) {
  if (this.latitude && this.longitude) {
    this.location = {
      type: "Point",
      coordinates: [this.longitude, this.latitude],
    };
  }
  next();
});

export default mongoose.model("Station",stationSchema);